File: leet_Design_0146_lru_cache___llist_map.cc
class LRUCache {
public:

    // Helper Node class
    class Node
    {

        public:
        int key;
        int val;
        Node *prev;
        Node *next;

        // constructor
        Node (int key, int val)
        {

            this->key = key;
            this->val = val;
            this->prev = NULL;
            this->next = NULL;
        }
    };

    // Helper -> For a circular linked list, needed dummy nodes
    Node    *head = new Node(0, 0);
    Node    *tail = new Node(0, 0);

    // Helper global tracker
    int Cap;
    unordered_map<int, Node *>  Dict;

    // helper fn :  remove . insert

    void    remove( Node *node )
    {

        Dict.erase( node->key );
        node->next->prev = node->prev;
        node->prev->next = node->next;
    }

    void    insert( Node *node )
    {

        node->next = head->next;
        head->next->prev = node;
        head->next = node;
        node->prev = head;

        Dict[ node->key ] = node;
    }

    // END of define


    LRUCache(int capacity)
    {

        this->Cap = capacity;
        head->next = tail;
        tail->prev = head;
    }
    
    int get(int key)
    {

        if ( Dict.find(key) == Dict.end() )
        {
            return (-1);
        }

        Node    *node = Dict[ key ];

        remove( node );
        insert( node );
        return ( node->val );
    }
    
    void put(int key, int value)
    {

        if ( Dict.find(key) != Dict.end() )
        {
            remove( Dict[key] );
        }
        if ( Dict.size() == this->Cap )
        {
            remove( tail->prev );
        }
        insert( new Node( key, value ));
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: leet_Design_0146_lru_cache___llist_map.ts
class Node {

    public key: number
    public val: number
    public prev: Node | null
    public next: Node | null

    constructor( key: number, val: number) {

        this.key = key
        this.val = val
        this.prev = null
        this.next = null
    }
}


class LRUCache {

    private head: Node
    private tail: Node
    private Cap: number
    private Dict: {
        [key: number]: Node
    }

    constructor(capacity: number) {

        this.Cap = capacity
        this.head = new Node(0, 0)
        this.tail = new Node(0, 0)
        this.head.next = this.tail
        this.tail.prev = this.head
        this.Dict = {}
    }

    private remove(node: Node) {

        delete this.Dict[node.key]

        if ( node.prev ) {
            node.prev.next = node.next
        }
        if ( node.next ) {
            node.next.prev = node.prev
        }
    }

    private insert(node: Node) {

        node.next = this.head.next
        this.head.next!.prev = node
        this.head.next = node
        node.prev = this.head

        this.Dict[node.key] = node
    }

    get(key: number): number {

        if ( !(key in this.Dict) ) {
            return -1
        }
        const node = this.Dict[key];
        this.remove(node)
        this.insert(node)

        return node.val
    }

    put(key: number, value: number): void {

        if (key in this.Dict) {
            this.remove(this.Dict[key])
        }
        if (Object.keys(this.Dict).length === this.Cap) {
            this.remove(this.tail.prev!)
        }
        this.insert(new Node(key, value))
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
